monai/networks/blocks/selfattention.py

```diff
@@ -62,11 +72,30 @@ def __init__(
         self.scale = self.head_dim**-0.5
         self.save_attn = save_attn
         self.att_mat = torch.Tensor()
+        self.rel_positional_embedding = (
+            get_rel_pos_embedding_layer(rel_pos_embedding, input_size, self.head_dim, self.num_heads)
+            if rel_pos_embedding is not None
+            else None
+        )
+        self.input_size = input_size
+
+    def forward(self, x: torch.Tensor):
+        """
+        Args:
+            x (torch.Tensor): input tensor. B x (s_dim_1 * ... * s_dim_n) x C
 
-    def forward(self, x):
+        Return:
+            torch.Tensor: B x (s_dim_1 * ... * s_dim_n) x C
+        """
         output = self.input_rearrange(self.qkv(x))
         q, k, v = output[0], output[1], output[2]
-        att_mat = (torch.einsum("blxd,blyd->blxy", q, k) * self.scale).softmax(dim=-1)
+        att_mat = torch.einsum("blxd,blyd->blxy", q, k) * self.scale
+
+        # apply relative positional embedding if defined
+        att_mat = self.rel_positional_embedding(x, att_mat, q) if self.rel_positional_embedding is not None else att_mat
```

Comment:
Agreed with Eric, we should consider how to refactor these blocks. Not sure whether these parts can be used together, if yes, perhaps we should add them in one block otherwise we may add a base class to always avoid large classes.
https://github.com/Project-MONAI/MONAI/blob/c61c6ac2d56af08fbbfb955324b8639e266a25db/monai/networks/nets/transformer.py#L116

Do you have some suggestions here?
cc @ericspod @marksgraham @vgrau98

Reply:
the use of `flash attention` does not look compatible with relative positional embeddings to me, as RPE requires access to the attention matrix, whilst flash attention calculates this behind the scenes. but I think we could just check if the user has selected both and throw a warning and/or deactive flash attention if that is the case. torch 2.0 now has a native implementation of flash attention [here](https://pytorch.org/docs/stable/generated/torch.nn.functional.scaled_dot_product_attention.html) so we should move to that when we can, too, rather than rely on an external library.

i believe the `causal` option should be compatible with RPE - we just set some of the attention matrix elements to `-inf` after adding the RPE to it. do you agree @vgrau98?

Reply:
I agree, we can have a check for Pytorch version and use the included function if version 2.0 or higher.

Reply:
i agree @marksgraham